Time sync howto (#563)
pupil-labs/pupil-docs: src/invisible/how-tos/data-collection-with-the-companion-app/achieve-super-precise-time-sync.md

Depending on the operating system, devices use different master clock servers to sync with. Ideally, all devices would sync to the exact same master clock to avoid small errors between different masters. Some operating systems allow to specify which server to use. Android is using the following ones depending on the region:

```
Asia
NTP_SERVER=asia.pool.ntp.org

Europe
NTP_SERVER=europe.pool.ntp.org

North America
NTP_SERVER=north-america.pool.ntp.org
```

### Improving Synchronization further
While an error of `<20 ms` is sufficient for most applications, some require even better synchronization. To improve further, you can estimate the exact offset between the clock used by Pupil Invisible and the external clock down to the millisecond.

This can be done using the `TimeOffsetEstimator` of the real-time API. Using the following code, you can estimate the offset between the Pupil Invisible clock and the clock of the host executing the code.

> 

Reviewer: Add dependency instruction
```suggestion
**Dependency**: `pip install "pip install pupil-labs-realtime-api>=1.1.0"`
```

---
description: How-To guide on achieving super-precise time sync using the real-time API.
---

# Achieve super-precise Time Sync

> For some applications, it is critical to accurately synchronize your Pupil Invisible device with another clock. That could be a second Pupil Invisible device, an external sensor or a computer you use for stimulus presentation. 

Reviewer 2: ```suggestion
For some applications, it is critical to accurately synchronize your Pupil Invisible device with another clock. That could be from a second Pupil Invisible device, an external sensor, or a computer you use for stimulus presentation. 
```

> Pupil Invisible provides UTC timestamps for all the data it generates, which makes it easy to sync the data to anything. Those timestamps are generated using the clock of the Companion device. However, all digital clocks suffer from drift, meaning that they are running slightly too fast or slow at all times. Over time this error accumulates and can lead to errors when comparing two clocks.

Reviewer 2: ```suggestion
Pupil Invisible provides UTC timestamps for all the data it generates, which makes it easy to sync the data to anything. Those timestamps are generated using the clock of the Companion device. However, digital clocks can suffer from drift, meaning that they sometimes run slightly too fast or slow. Over time this error accumulates and can lead to errors when comparing two clocks.
```

This can be done using the `TimeOffsetEstimator` of the real-time API. Using the following code, you can estimate the offset between the Pupil Invisible clock and the clock of the host executing the code.
::: tip
**Dependency**: `pip install "pupil-labs-realtime-api>=1.1.0"`
:::
```python
from pupil_labs.realtime_api.simple import discover_one_device

# Look for devices. Returns as soon as it has found the first device.
print("Looking for the next best device...")
device = discover_one_device(max_search_duration_seconds=10)
if device is None:
    raise SystemExit("No device found.")

estimate = device.estimate_time_offset()
if estimate is None:
    device.close()
    raise SystemExit("Pupil Companion app is too old")

print(f"Mean time offset: {estimate.time_offset_ms.mean} ms")
print(f"Mean roundtrip duration: {estimate.roundtrip_duration_ms.mean} ms")

device.close()
```

Using continuous offset estimates like this, you can precisely compensate for clock drifts by correcting the respective timestamps.

> When using a wifi connection, the speed of data transfer can vary at times, leading to inaccuracies in the estimation. If precise synchronization down to a few milliseconds is required, it is recommended to use an ethernet connection. You can connect an ethernet cable to the phone using an appropriate USB-C adapter or hub.

Reviewer 2: I feel like this paragraph should be moved to a more prominent location, since the whole super millisecond accuracy kind of depends on a wired connection, unless I'm misunderstanding.

I think a good location would be perhaps as a tip/aside after this sentence: 
"This can be done using the TimeOffsetEstimator of the real-time API."

Author: TBH I am not sure how large the influence of the wifi is. The `TimeOffsetEstimator` is running a series of measurements and averages across them, which should largely alleviate inaccuracies from noise in the wifi speed.
I *think* the accuracy just decreases from +-1 ms to +-3-4 ms max, but we might want to measure this to be sure.

Reviewer 2: Ah ok. In that case, I think it's worth measuring it so the reader has a rough idea of the impact. Otherwise, the statement is a bit nebulous imo

> When using a wifi connection, the speed of data transfer can vary at times, leading to inaccuracies in the estimation. If precise synchronization down to a few milliseconds is required, it is recommended to use an ethernet connection. You can connect an ethernet cable to the phone using an appropriate USB-C adapter or hub.

Reviewer 2: Looks good, @marc-tonsen. I've left some comments + suggested edits.

Using continuous offset estimates like this, you can precisely compensate for clock drifts by correcting the respective timestamps.

> When using a wifi connection, the speed of data transfer can vary at times, leading to inaccuracies in the estimation. If precise synchronization down to a few milliseconds is required, it is recommended to use an ethernet connection. You can connect an ethernet cable to the phone using an appropriate USB-C adapter or hub.

Reviewer 3: @marc-tonsen maybe it is worth specifying the hub is only available for One Plus 8 and 8t?

Author: Agreed! I wonder if instead of putting this in here, we should link out to a dedicated page on using HUBs. Maybe as part of the "Glasses & Companion Device" section?

Reviewer 3: makes sense, as they might use it for charging only, and not for the ethernet capabilities

Author: I have added an according section and linked to it. Please check it out!
